### ISSUE TITLE
Bug - On dev branch - ubuntu 20 - open / save file modal will freeze and crash app

### ISSUE THREAD

**Issue by olup** [tauri-apps/tao#81]
This is a bug

Using Tauri with patched Tao (latest dev branch) on ubuntu 20.10.

Trying to open a file modal (to open or save) will try to open something - it shows in the task bar - but no window is actually drawn to the screen. It hangs and eventually crash : "App is not responding / Force quit / Wait"

**Comment by lemarier** [comment]
Thanks for reporting.

We'll do our tests and keep you updated.

**Comment by olup** [comment]
Just wanted to add that I just tried it on macos - no issue. I'll test on windows too.

**Comment by lucasfernog** [comment]
Should be a Linux-only issue. I'll help track this.

**Comment by lucasfernog** [comment]
The problem started after tauri-apps/tao#68, ControlFlow::Wait and ControlFlow::Poll are not working. I guess we'll need your help here @wusyong

**Comment by wusyong** [comment]
Sure, I can look into it. Is there an example (on tao) I could test?

**Comment by lucasfernog** [comment]
Not really, we just run "cargo run --bin api" on the tauri repo and open the dialog tab.

**Comment by wusyong** [comment]
Okay, I can confirm this is because of the new control flow since it now also runs on the main loop aside with `gtk::main_iteration`.
Unfortunately, I don't think we could change the implementation. Because it is the only way to work with all control flow and many other event variants afaik.
The workaround I think could spawn a glib task (`Maincontext::spawn` or `invoke`) when calling dialog.
Sine tao already pushes the thread default into the Maincontext, you should be able to get the Maincontext and use it right away.
```rust
let context = glib::MainContext::default();
context.invoke(|| {...});
```

**Comment by olup** [comment]
Hi, any progress on this ? Should this issue be moved to tauri repo ?

**Comment by lemarier** [comment]
> Hi, any progress on this ? Should this issue be moved to tauri repo ?

Let's see what @lucasfernog has to say about that.

**Comment by lemarier** [comment]
> Okay, I can confirm this is because of the new control flow since it now also runs on the main loop aside with `gtk::main_iteration`.
> Unfortunately, I don't think we could change the implementation. Because it is the only way to work with all control flow and many other event variants afaik.
> The workaround I think could spawn a glib task (`Maincontext::spawn` or `invoke`) when calling dialog.
> Sine tao already pushes the thread default into the Maincontext, you should be able to get the Maincontext and use it right away.
> 
> ```rust
> let context = glib::MainContext::default();
> context.invoke(|| {...});
> ```

Tried to test if it can work by patching the `message` in Tauri, and I still got the same issue;
(disabling `rfd` work, so the issue comes from there)

```rust
  let context = glib::MainContext::default();
  context.invoke(move || {
    rfd::MessageDialog::new()
    .set_title(title.as_ref())
    .set_description(message.as_ref())
    .set_buttons(rfd::MessageButtons::Ok)
    .set_level(rfd::MessageLevel::Info)
    .show();
  });
```

I'll check `rfd` how they work... 
Edit; they use `std::thread::spawn` into the GTK integration that causes the crash.
